src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Windows/KeyAgreeRecipientInfoPalWindows.cs

```diff
@@ -0,0 +1,188 @@
+                    throw new InvalidOperationException(SR.Cryptography_Cms_Key_Agree_Date_Not_Available);
+
+                unsafe
+                {
+                    return WithCmsgCmsRecipientInfo<DateTime>(
+                        delegate (CMSG_KEY_AGREE_RECIPIENT_INFO* recipient)
+                        {
+                            CMSG_RECIPIENT_ENCRYPTED_KEY_INFO* pEncryptedKeyInfo = recipient->rgpRecipientEncryptedKeys[SubIndex];
+                            long date = (((long)(uint)(pEncryptedKeyInfo->Date.dwHighDateTime)) << 32) | ((long)(uint)(pEncryptedKeyInfo->Date.dwLowDateTime));
+                            DateTime dateTime = DateTime.FromFileTimeUtc(date);
+                            return dateTime;
+                        });
+                }
+            }
+        }
+
+        public sealed override CryptographicAttributeObject OtherKeyAttribute
+        {
+            get
+            {
+                if (RecipientIdentifier.Type != SubjectIdentifierType.SubjectKeyIdentifier)
+                    throw new InvalidOperationException(SR.Cryptography_Cms_Key_Agree_Date_Not_Available);
+
+                unsafe
+                {
+                    return WithCmsgCmsRecipientInfo<CryptographicAttributeObject>(
+                        delegate (CMSG_KEY_AGREE_RECIPIENT_INFO* recipient)
+                        {
+                            CMSG_RECIPIENT_ENCRYPTED_KEY_INFO* pEncryptedKeyInfo = recipient->rgpRecipientEncryptedKeys[SubIndex];
+                            CRYPT_ATTRIBUTE_TYPE_VALUE* pCryptAttributeTypeValue = pEncryptedKeyInfo->pOtherAttr;
+                            if (pCryptAttributeTypeValue == null)
+                                return null;
+
+                            string oidValue = pCryptAttributeTypeValue->pszObjId.ToStringAnsi();
+                            Oid oid = Oid.FromOidValue(oidValue, OidGroup.All);
+                            byte[] rawData = pCryptAttributeTypeValue->Value.ToByteArray();
+                            Pkcs9AttributeObject pkcs9AttributeObject = new Pkcs9AttributeObject(oid, rawData);
+                            AsnEncodedDataCollection values = new AsnEncodedDataCollection(pkcs9AttributeObject);
+                            return new CryptographicAttributeObject(oid, values);
+                        });
+                }
+            }
+        }
+
+        internal int Index { get; }
+        internal int SubIndex { get; }
+
+        // Provides access to the native CMSG_KEY_TRANS_RECIPIENT_INFO* structure. This helper is structured as taking a delegate to 
+        // help avoid the easy trap of forgetting to prevent the underlying memory block from being GC'd early.
+        internal T WithCmsgCmsRecipientInfo<T>(KeyAgreeReceiver<T> receiver)
+        {
+            unsafe
+            {
+                CMSG_CMS_RECIPIENT_INFO* pRecipientInfo = (CMSG_CMS_RECIPIENT_INFO*)(_pCmsgCmsRecipientInfoMemory.DangerousGetHandle());
```

Comment:
I’m not completely opposed to using the safer pattern as a belt-and-suspenders things, but RecipientInfoPal’s have no Dispose() method and can never realistically benefit from one because the RecipientInfo (API type) that contains them has no Dispose method either. (And it would look weird to add one since RecipientInfo’s are supposed to be “just data”.)